add a reinplacing FX pass

bdhirsh opened the pull request:
Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.

There are some examples in the test file, but I think staring at an example of where re-inplacing is/isn't allowed to run is helpful:
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input,
    // because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
    // second add() is NOT safe to re-inplace, because:
    // (1) a and tmp1 are aliased. Note that they weren't aliased in the original program,
             but they are now that we've done some re-inplacing.
    // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```


Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80897
* #82009
* #82008
* #82007

bdhirsh committed:
add a reinplacing FX pass

[ghstack-poisoned]

bdhirsh committed:
Update on "add a reinplacing FX pass"

Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input, because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
   // second add() is NOT safe to re-inplace, because:
   // (1) a and tmp1 are aliased
   // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```





[ghstack-poisoned]

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+    # Both tensors are views, of the same original base tensor
+    if a._base is not None and b._base is not None and a._base is b._base:
+        return True
+    return False
```
`_base` is not always sound; e.g. inference mode tensors won't have base set

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+                kwargs = load_arg(node.kwargs)
+                result = getattr(self_obj, node.target)(*args, **kwargs)
+            elif node.op == 'call_module':
+                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))
```
The boilerplate here can get eliminated with Interpreter

SherlockNoMad commented:
> I think this is ready for review - from my basic tests so far I think it works, but I'm working on trying to set up the torchbench suite locally so I can run all of the models through functionalization + re-inplacing and surface other stuff.
> 
> I also haven't written a major FX transform before, so welcoming any feedback on best practices that I'm not following.
> 
> cc. @SherlockNoMad in case you're interested - once this is ready I think we'll want to hook this up to E2E primtorch + NVFuser benchmarks?

Yes, I added this as a todo item.

SherlockNoMad commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+    return names_match and arg_types_match
+
+@compatibility(is_backward_compatible=True)
+def reinplace(gm: fx.GraphModule, *sample_args, **sample_kwargs):
```
I am wondering if sample inputs are needed for this pass... 
My first intuition is that buffer reuse planning can be done by just examining the graph's topological structure, without a concrete input. 

Take following as an example:
b = a.add(...)
c = b.add(...)

b can reuse a's storage if it meets the following condition:
- a's user count is 1 
- a's buffer can be overwritten (this cover the case where a is a graph input)

ezyang replied:
I guess one can always come up with annoying cases where ownership is unclear, e.g., x.reshape, though I suppose  you could just conservatively assume that you don't have ownership in that case.

SherlockNoMad commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+    inputs to the program.
+
+    There is one exception though: if "a" is copied to at any point later in the program,
+    e.g. `a.copy_(...)`, then we are free to re-use a's buffer any time
```
I am wondering can we eliminate such case by making it a requirement that input graph_module is purely functional, so you cannot have any copy_ node in the graph. 

This requirement is achieved by upstream functionalization pass, which ensures that the output graph is functional. 

In a sense, functionalization and reinplace are two gates to enter and exit the functional world.

SherlockNoMad commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+def reinplace(gm: fx.GraphModule, *sample_args, **sample_kwargs):
+    """
+    Given an fx.GraphModule, modifies it to perform "reinplacing".
+    We look for out-of-place op call sites like `b = a.add(...)`,
```
What's assumption of IR for this pass? 
`b = a.add(...)` seems to imply that it's working on torch IR, but it should also work for aten IR... 

or any functional IR?

SherlockNoMad commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+from typing import Iterable
+import itertools
+
+__all__ = ['reinplace']
```
This is the PassBase interface we wish to standardize... Appreciate any feedback 

https://github.com/pytorch/pytorch/blob/729c56d1363a61ec5324743a6a3a3098a084153e/torch/fx/passes/infra/pass_base.py

ezyang replied:
I guess postcondition would be a good place to add the "check that the fake tensor annotations are still up to date after the transform"

bdhirsh committed:
Update on "add a reinplacing FX pass"

Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.

There are some examples in the test file, but I think staring at an example of where re-inplacing is/isn't allowed to run is helpful:
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input,
    // because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
    // second add() is NOT safe to re-inplace, because:
    // (1) a and tmp1 are aliased. Note that they weren't aliased in the original program,
             but they are now that we've done some re-inplacing.
    // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```





[ghstack-poisoned]

bdhirsh commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+    For a given variable name "x", alias_map[x] gives you a set
+    of all variables in the graph that alias with x (including x itself).
+    """
+    def __init__(self, mod):
```
thanks, that was a nice template to base my changes off of

SherlockNoMad commented:
> updated: the main changes are that I turned all of the pre-processing bookkeeping steps into a single `fx.Interpreter` pass. I still do the actual inplace-ing as a graph transforms though, because (I think) that I need to.
> 
> The reason is that as I'm in-placing a node in the graph, that requires updating all of my existing data structures. Doing so requires visiting future nodes in the graph. I can't re-inplace any other nodes though until those future nodes have been visited.

If the re-inplacing implementation is tied to the execution order (as you mentioned "future" node), then I am concerned that if "recompile" returns a different execution order, the planed inplaced might break.

bdhirsh commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,303 @@
+@compatibility(is_backward_compatible=True)
+def reinplace(gm: fx.GraphModule, *sample_args, **sample_kwargs):
+    """
+    Given an fx.GraphModule, modifies it to perform "reinplacing".
```
hmmm, that is one thing that I was thinking about.

This pass only does the first thing (inplace planning), although I could imagine squeezing out even more memory savings by doing memory planning, like what the PyE team is working on - convert all ops into out= variants, turn all tensors into offsets into a giant buffer that are calculated for optimal re-use.

I was wondering - do we actually want that as an always-on feature in our E2E stack (in the same way that functionalization + re-inplacing will be "always on" when running the E2E stack)?

It seems like something you may or may not want to use, depending on the backend: if you're using NVFuser, would NVFuser know how to handle a bunch of out= operations properly? And if you're using XLA, well... it already does it's own memory planning optimizations, right?

bdhirsh commented:
@SherlockNoMad and I talked offline - summarizing our conversation (lmk if I got anything wrong):

(1) re-inplacing "enforcing an execution order"

The pass currently in this PR does re-inplacing more aggressively, in a way that's meant to fully get back all of the perf lost through functionalization. As a result, it's tied to the execution order (it involves looking ahead in the graph, and is also more complicated). Simplest example:
```
def foo(a):
    b = a.add(...)
    c = a.add(...)
    a.copy_(b)
```
The PR today will re-inplace the first op but not the second, because that's exactly what'll get back performance parity after running functionalization.

But because the first two lines in the above program are independent, that also means that the current PR will force an execution order: after re-inplacing, line 1 needs to run before line 2. You could imagine other optimization passes that might want to look at multiple execution orders and pick one.

Will we want to run other (potentially backend-specific) optimization passes that change the execution order, that are somehow forced to run after re-inplacing? With NVFuser and cuda graphs at least, I think the answer is no. For NVFuser specifically, we want to first run the pass to create fusion groups, and then at the end perform re-inplacing (the re-inplacing pass will just ignore fusion groups because it won't know how to re-inplace them). But if we encounter a backend in the future where this is problematic, it doesn't seem like too big of a lift to use a slightly different re-inplacing pass that works better for them

(2) re-inplacing requiring changing the inputs of other nodes

When replacing `b = a.add(...)` with `b = a.add_(...)`, I also update all future nodes in the graph that take in `b` as an input, and swap the input with `a`. That definitely makes the pass more complicated, and it's also unnecessary for `add_()` (and most other inplace ops), because `add_()` returns its input (so `b` and `a` are both valid variables, that are really the same tensor.

I wrote it that way because technically, our ATen inplace schemas are not required to return their `self` arguments. This is only actually true today for inplace ops that take in a `TensorList` input (today this is most of the `_foreach_{op}_` operators). So... I guess the tradeoff here is that we chose to have `foo_(Tensor[](a!) self), ...)` schemas in ATen return `void` (which is more intuitive because you can't method chain on tensor lists), but as a result the logic in this pass is a bit more complicated.

bdhirsh commented:
(fyi the unit tests that I wrote all pass, but was able to run torch dynamo's torchbench suite locally and hit some issues that I'm starting to look at. I think the PR is in a reasonable state for review though)

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+        self.alias_map = defaultdict(set)
+        # For a given variable, tells you the names of all nodes in the program that use it.
+        # (including the index of that node, so we know where in the program it is)
+        self.names_to_used_nodes = defaultdict(list)
```
It's funny to see this because in TorchScript IR, uses information (what nodes use a node) is so important that it is baked into the IR representation and must be kept up to date haha. FX isn't like that though lol.

ezyang replied:
list is pretty non-descriptive, so a type annotation or a comment saying how exactly the node + index is represented would help here

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+            return super().run_node(node)
+
+
+    def __call__(self, *args, **kwargs):
```
drop the kwargs here

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+    """
+
+
+    sample_args = [a.to(device='meta') if isinstance(a, torch.Tensor) else a for a in sample_args]
```
OK so after the Friday discussion we agreed that we should be using the fake tensor stored metadata to get this information, rather than recomputing it on the fly in the Interpreter. https://github.com/pytorch/pytorch/pull/80566/files#diff-10df4b56573d38527c0d4485d31ac62c3aa4d76077f8d2066eb71abb7c2bc0c7R101-R128 is an example of how to do it in the new style but you'll need https://github.com/pytorch/pytorch/pull/80566/files#diff-342fa21508bcf89a6b85251ce90db2ea534a1bb06f72441a4714569f47d1d9c9R32-R42 landed somewhere first so you can actually populate these nodes (actually, we should probably make proxy tensor populate this as we are tracing; cc @Chillee ). Sorry about the API whiplash.

Btw if you still wanted to do this as is you should create a fake tensor mode and then use the converter there to convert to fake tensors, as that will maintain aliasing relationships in the inputs, which this code does not do correctly.

ezyang reviewed (approved):
let's land this for now and we can iterate

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+        # When you call `x.copy_(y)`, you overwrite x's data.
+        # This stores the name of the node (if any) where a given variable in the program is overwritten.
+        # We need to track this information for program inputs
+        self.last_node_where_variable_is_overwritten = defaultdict(lambda: -1)
```
So if we setup the tiered representation, where `copy_` is kept out of the IR, then we actually wouldn't have to deal with this; instead, we would have a separate optimization pass which "collapses" the tiered rep.

Actually this makes me wonder if we should represent the tiering as just a nested module. So you have an inner GraphModule which is purely functional, and then the outer container GraphModule which has the copies. The alternative was to have an extra epilogue module but I wonder if explicitly having the nesting makes it clearer. One downside is that all optimization passes would need to know to "look past" the containing module into the functional inside, so maybe this is not so great.

ezyang replied:
Another possibility is we create a GraphModule subclass where the graph is functional, but it has some extra metadata saying what `copy_` to do and we override the codegen to insert these copies when you actually generate the code for the module. 

I wonder if @Chillee has an opinion.

bdhirsh replied:
It seems to me like in a lot of cases, re-inplacing will basically turn the side-car `copy_()` module/graph into a no-op (by folding it back into a reinplace on the input). I guess the main case where this wouldn't happen is if we're using a fuser and the fuser already added the op into a fusion group.

I don't have a strong opinion, but it feels easy / clean just to keep the `copy_()`'s in a separate module, and leave it up the re-inplacing to optionally fold them back into the graph:
```
# output represents the newly inplaced module, and the new version of the `copy_()` module (where some of those copies may have been removed, and folded into the first module).
def reinplace(functional_graph: fx.GraphModule, input_mutation_graph: fx.GraphModule) -> Tuple[fx.GraphModule, fx.GraphModule]
```

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+                if isinstance(a, fx.node.Node):
+                    # We map to the actual fx.Node object, because later we might need to
+                    # modify the arguments that this node is called with
+                    self.names_to_used_nodes[a.name].append([self.node_counter, node])
```
It's probably safer to use tree_map for this; as written this code can only handle Tensor and list of Tensor, if a Tensor pops up anywhere else it will be missed.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+            # Update alias map + input names when we see a placeholder
+            r = super().placeholder(node.target, args, kwargs)
+            self.input_names.append(node.name)
+            self.alias_map[StorageWeakRef(r.storage()).cdata].add(node.name)
```
it should be possible to index just on the StorageWeakRef directly, although I remember you telling me that this didn't work...

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+            for inpt_name in aliased_inputs:
+                last_node_where_input_alias_is_overwritten[inpt_name] = copy_node_idx
+
+        return self.input_names, alias_map, self.names_to_used_nodes, last_node_where_input_alias_is_overwritten
```
@Chillee so, if I understand you correctly, idiomatic way to implement this would be to FakeTensorProp and then use the fake tensor annotations on nodes instead of reinterpreting, right?

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+    # for other arbitrary operators).
+    assert len(inplace_overloads_with_matching_schemas) == 1
+    inplace_op = inplace_overloads_with_matching_schemas[0]
+    return inplace_op
```
Can we use torchgen to do this logic instead?

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+
+                input_alias_names = set()
+                for inpt_name in input_names:
+                    input_alias_names |= alias_map[inpt_name]
```
isn't this always the same every loop iteration

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,257 @@
+                    # A lot of the extra below is because we have several data structures
+                    # That we need to update when we change the nodes' arguments:
+                    # - names_to_used_nodes (some nodes now take in different variables as arguments)
+                    # - alias_map (updating a view nodes' arguments will change with variables alias)
```
This part definitely feels like it would be easier if you were rebuilding the graph instead of mutating. You'd maintain a mapping env and apply it as you progressed further

ezyang commented:
I think I actually want something more ambitious than reinplacing: I want the inverse of functionalization. In particular:

* I want views to be restored
* Suppose the original program had a mutation on a view, which functionalized into this operation being performed multiple times. I want those operations to get coalesced back into a single mutation

As a testing strategy for this, it would be nice to take all the preexisting functionalization tests, and run invert functionalization on the result of functionalization and show that you get something like the original.

Chillee commented:
@ezyang I've also said this, but I want something that's *structurally* the inverse of functionalization. So, maybe functionalization produces some kind of certificate about what it's functionalizing, and we use that to perform the inverse of functionalization.

bdhirsh committed:
Update on "add a reinplacing FX pass"

Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.

There are some examples in the test file, but I think staring at an example of where re-inplacing is/isn't allowed to run is helpful:
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input,
    // because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
    // second add() is NOT safe to re-inplace, because:
    // (1) a and tmp1 are aliased. Note that they weren't aliased in the original program,
             but they are now that we've done some re-inplacing.
    // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```





[ghstack-poisoned]

ezyang commented:
holler when you want a review, I'll be looking for an updated algo description

bdhirsh committed:
Update on "add a reinplacing FX pass"

Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.

There are some examples in the test file, but I think staring at an example of where re-inplacing is/isn't allowed to run is helpful:
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input,
    // because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
    // second add() is NOT safe to re-inplace, because:
    // (1) a and tmp1 are aliased. Note that they weren't aliased in the original program,
             but they are now that we've done some re-inplacing.
    // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```





[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -205,7 +211,7 @@ void FunctionalTensorWrapper::replace_(const Tensor& other) {
     // .to() should not re-entrantly go through functionalization.
     at::AutoDispatchSkipFunctionalize guard;
-    value_ = value_.to(c10::TensorOptions().dtype(dtype()).layout(layout()));
+    value_ = at::_to_copy(value_, c10::TensorOptions().dtype(dtype()).layout(layout()));
```
ok this was sneaky, but I noticed that `to.dtype_layout` was showing up directly in some of my FX graphs, which made the reinplace pass unhappy (it doesn't expect to see any composite ops, since e.g. `to()` is a "maybe view" operator).

Why does that happen? autograd has already ran by the time we get to this code (in my crappy implementation of `_functionalize()` in the test suite), so that prevents the composite kernel for `to()` from being called here.

Putting `_to_copy()` into the graph is what we want. I guess this would also get fixed by registering composite kernels directly to the `Python` key though..

ezyang replied:
This looks good, but you should have a comment in the code here

bdhirsh commented:
@ezyang this is ready for another round. Major changes:

(1) Updated the pass to do the more complicated thing to re-inplace `{view}_scatter` calls. I added a more detailed bit on the algorithm in the doc-string.

(2) Updated `test_functionalization` to heavily test the rein place pass. Every test example will now run both functionalization and reinplacing and assert that the inputs + outputs match. I also printed the reinplaced code with expect tests in a few places.

(3) Fixes some bugs that came out of the above (lower in the stack)

(4) I removed the extra optimization for handling `copy_()` reinplacing for now. I think it makes more sense to first land a version of "moving program input mutations into a separate submodule", and then figure out how to fold that back into the reinplace pass.

ezyang commented on test/test_fx_reinplace_pass.py:
```diff
@@ -0,0 +1,114 @@
+            return b
+
+        inpt = torch.ones(2)
+        f2 = reinplace(make_fx(f)(inpt), inpt)
```
If we do https://github.com/pytorch/pytorch/issues/82044 it shouldn't be necessary to feed inputs into `reinplace` as it should be able to just rely on the fake tensors stashed on the FX IR itself.

ezyang commented on test/test_functionalization.py:
```diff
@@ -104,10 +118,27 @@ def f(x):
+    view_default_1 = torch.ops.aten.view.default(add_tensor, [4, 2])
+    mul_tensor = torch.ops.aten.mul.Tensor(view_default_1, view_default_1)
+    copy__default = torch.ops.aten.copy_.default(a_1, view_default_1);  a_1 = view_default_1 = None
     return add_tensor
```
These expects are going to get a bit tiresome to audit by hand. Can we come up with some invariants to test? For example, I'd expect functionalize . reinplace (or vice versa) to be idempotent (maybe after an initial reinplace to deal with programs that aren't totally reinplaced as necessary). You'll need a way to test graph equivalence but IMO it's worth it.

ezyang replied:
Another simple thing to test: do the graphs give the same outputs given the same inputs

bdhirsh replied:
I'm pretty sure that... Once we (a) add the optimization for `copy_()`, and (b) add an extra pass / bit of logic to remove duplicate views, then the output of `reinplace` should be the same as the original like you're saying. Once those two changes are in, I'll test out removing these extra expect tests

ezyang commented on test/test_functionalization.py:
```diff
@@ -104,10 +118,27 @@ def f(x):
+    add_tensor = torch.ops.aten.add.Tensor(view_default, empty);  view_default = empty = None
+    view_default_1 = torch.ops.aten.view.default(add_tensor, [4, 2])
+    mul_tensor = torch.ops.aten.mul.Tensor(view_default_1, view_default_1)
+    copy__default = torch.ops.aten.copy_.default(a_1, view_default_1);  a_1 = view_default_1 = None
```
It looks like this didn't do a very good job of propagating the mutations directly to the input 🤔

bdhirsh replied:
Hmmm what do you mean?

(Once we add the optimization for `copy_()`, re-inplacing should know how to re-inplace the `add()` in this example and remove the `copy_()` - I just figured it's not worth adding that optimization until we first add the logic to split off input mutations into their own submodule).

ezyang replied:
Oh oops, you are right, you haven't implemented this yet

ezyang commented on test/test_functionalization.py:
```diff
@@ -134,6 +165,21 @@ def forward(self, a_1):
+    view_default = torch.ops.aten.view.default(a_1, [4, 2]);  a_1 = None
+    empty_1 = torch.ops.aten.empty.SymInt([], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
+    add_tensor = torch.ops.aten.add.Tensor(view_default, empty);  view_default = empty = None
+    mul_tensor = torch.ops.aten.mul_.Tensor(add_tensor, add_tensor)
```
Hmm, we probably need to explicitly forbid this. Inplace with aliasing arguments often does not work with eager kernels. Not worth the potential correctness headache.

bdhirsh replied:
bleh good point. I'll patch this before merging - "if the mutable argument shows up multiple times in the arg list, don't re-inplace"

ezyang commented on test/test_functionalization.py:
```diff
@@ -150,6 +196,20 @@ def f(x):
 
+def forward(self, a_1):
+    empty = torch.ops.aten.empty.SymInt([4], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
+    empty_1 = torch.ops.aten.empty.SymInt([4], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
```
These are dead, but once we run reinplacing we cannot run standard DCE because it will delete too many nodes. I guess we should make a mutation sound DCE.

ezyang replied:
Actually, this is the functional graph. This doesn't seem that interesting from a functionalization perspective lol; although it suggests we should run DCE after functionalizing.

bdhirsh replied:
oh yep, agreed - and we already run DCE as part of partitioning today, so these will get removed.

ezyang commented on test/test_functionalization.py:
```diff
@@ -165,7 +225,38 @@ def f(x):
+    view_default = torch.ops.aten.view.default(lift_fresh, [-1]);  lift_fresh = None
+    add_tensor = torch.ops.aten.add_.Tensor(view_default, 1)
+    view_default_1 = torch.ops.aten.view.default(view_default, [3]);  view_default = None
+    return view_default_1
```
as you said, no built in view removal

ezyang commented on test/test_functionalization.py:
```diff
@@ -185,9 +276,25 @@ def forward(self, a_1):
     empty = torch.ops.aten.empty.memory_format([4, 2], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
     fill_scalar = torch.ops.aten.fill.Scalar(empty, 1.0);  empty = None
     view_copy_default = torch.ops.aten.view_copy.default(a_1, [4, 2])
```
I stopped auditing here

ezyang commented on test/test_functionalization.py:
```diff
@@ -744,6 +1044,21 @@ def forward(self, a_1):
+    add_tensor_1 = torch.ops.aten.add.Tensor(view_default_1, 1)
+    return (view_default_1, add_tensor_1)
     """)
 
```
I noticed I didn't see any tests in the set here for things like:

```
x = torch.zeros(10)
x[5] = 1
return x
```

bdhirsh replied:
Good catch :D this breaks something - it looks like `FakeTensor` isn't handling the scalar-tensor conversion here well, when the `1` gets converted to a (non-fake) scalar-tensor.

ezyang replied:
Oh I feel like this is an unrelated bug that @eellison is fixing right now, see https://github.com/pytorch/pytorch/pull/81927

Figure out some other way to do this for now. How about `x[5].fill_(1)`

bdhirsh replied:
ah nice

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+        There are a few caveats to this, explained in more detail below:
+        (a) If "a" is used later as an argument to a view op, that is okay.
+            It's only a problem if "a" (or that view) is later passed
+            into a normal operator, or if it is returned as the program output.
```
If the view is never passed into a normal operator, doesn't that mean it's dead? So you can simplify condition (a) if you DCE before reinplacing.

bdhirsh replied:
Yep that's true - what's a little annoying is that you might have a chain of a view op + a scatter op that both use "a" later in the graph, like:
```
b = a.add(...)
a_view = a.view(...)
a_select_scatter = select_scatter(a_view, ....)
```

In that case, both the `a_view` and `a_select_scatter` nodes show up as "nodes that use an alias of a later in the graph". The logic I have currently will recognize the `a_select_scatter` node as being ok to remove, but not the `a_view` node. So I figured it'd be simpler just to ignore any view ops when figuring out which nodes can prevent an op from being reinplaced.

I guess another option is that when we detect that the `a_select_scatter` node can be removed, we include it + it's entire view chain.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+            return a_updated
+
+        And "re-inplacing" is on the hook for figuring out how to remove
+        the expensive diagonal_scatter call.
```
"Ordinarily, we would not be able to reinplace the fill, because b aliases with a which is used by the diagonal_scatter call"

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+        And "re-inplacing" is on the hook for figuring out how to remove
+        the expensive diagonal_scatter call.
+
+        So, for every `alias in alias_set(a)`, instead of checking
```
do you mean b?

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+                      diagonal -> diagonal_scatter
+                      slice -> slice_scatter
+                      select -> select_scatter
+                      as_strided -> as_strided
```
orly

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+    (5) Any "view_inverse/scatter" nodes that were identified as "it's ok to ignore them"
+        during step (3) get manually deleted from the graph.
+        Their outputs are no longer used, so technically standard DCE would be able
+        to do this, but we can no longer run FX's DCE pass now that we mave mutable
```
have

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+# n.meta['node_usages']: List[List[int, Node]]
+#   A list of all nodes that take in the current node as an input
+#   Each element is actually an (int, Node) pair, specifying the index of the node in the graph.
+# n.meta['view_of']: Node
```
@SherlockNoMad informed me recently that FX node has a "users" property that lists node users. However, this doesn't have index information, so maybe you still need this

bdhirsh replied:
oh yep this works, thanks. I ended up storing the indexing info on the node's `meta` field directly instead

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+#   If the current node n is a view of some base tensor, the 'view_of' field tells us which
+#   view node was used to generate the current node (a view tensor).
+#   This information actually makes `fake_result` redundant, but we can use `fake_result`
+#   to sanity check that our aliasing information is correct.
```
Hmmmmm... technically, should `fake_result` also have recorded view metadata that also records this? (Not just strides, but the view function chain that we use to reconstruct views.) I admit it is probably not in a form that is easily accessible from Python.

bdhirsh replied:
Is that the autograd view metadata? Or something else?

ezyang replied:
View metadata

bdhirsh replied:
Hmm if this is the `DifferentiableViewMeta` in autograd, then don't we need to worry about inference mode again? (I guess we could just temporarily disable it internally while tracing).

Or are you suggesting to try to re-use the `ViewMeta` chains constructed from functionalization? hmm.


Either way I might leave this as a followup :)

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+                # modify the arguments that this node is called with
+                x.meta['node_usages'].append(node)
+
+        tree_map(_update_node_usage, list(itertools.chain(node_args, node.kwargs.values())))
```
tree map can see into dicts, so you don't need to actually chain this, `(node_args, node.kwargs)` would work just as well

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+            # And we'd like to set:
+            #    getitem.meta['view_of'] = x_1
+            #    getitem1.meta['view_of'] = x_1
+            if node.target is _operator.getitem:
```
elif right?

bdhirsh replied:
they should be mutually exclusive anyway but yep agreed

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+            #    %getitem_1 : [#users=1] = call_function[target=operator.getitem](args = (%split_tensor, 1), kwargs = {})
+            # And we'd like to set:
+            #    getitem.meta['view_of'] = x_1
+            #    getitem1.meta['view_of'] = x_1
```
nit getitem_1

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+        if len(schema.arguments) > 0:
+            first_arg = schema.arguments[0]
+            # check if op is a view
+            return first_arg.alias_info is not None and not first_arg.alias_info.is_write
```
@anjali411 seems like a tag opportunity. In particular this test is a bit wonky because hypothetically someone could define a view where the aliased argument isn't the first one, and that would silently break the invariants here.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+    return None
+
+def _can_reuse_tensor_memory(self_arg, *, view_of_self: Optional[torch.fx.node.Node] = None):
+    nodes_used = self_arg.meta['node_usages'] if 'node_usages' in self_arg.meta else []
```
is it possible for this field to not be set?

bdhirsh replied:
oops, unused function that I forgot to delete

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+    nodes_used = self_arg.meta['node_usages'] if 'node_usages' in self_arg.meta else []
+    # If the current tensor argument is used as an input to any ops later in the graph,
+    # then we can't re-use its memory
+    if any(_ for n_idx, n in nodes_used if n_idx > idx):
```
What does this do?

bdhirsh replied:
(unused function I forgot to delete)

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+#    a2 = a1[0]
+#    b1 = x.view(2, 4)
+#    b2 = b1[0]
+# Here, b2 and a2 are matching view tensors, because their chain of views is identical.
```
if the chain of views is not identical, but the strides are identical, would it make sense for these to count as matching views?

bdhirsh replied:
I spent some time thinking about this and I'm tempted to save as a follow-up lol. I think you're probably right, and it could simplify this code. It would probably also let us re-inplace any `*_scatter` ops that users use directly in their program, not just from functionalization. I'm a bit worried about false positives though, and making sure we don't accidentally re-inplace something we shouldn't.

bdhirsh replied:
Ok, I think I have an even simpler version of the "remove scatter ops" logic. Written below:
```
# Given an op that we're trying to re-inplace, "b = foo(a)",
# And given a {view}_scatter op that shows up later in the graph, "y = {view}_scatter(base, x, args...)"
# Then re-inplacing `foo()` would allow us to remove the `{view}_scatter` op entirely, IF:
# (1) "base" is in alias_set(a)
# (2) "x" is in alias_set(b)
```

There's a caveat though: If you have a chain of `{view}_scatter` calls that feed into each other, where the correct output would be to remove all of them, it's not enough to check that `x in alias_set(b)`. "x" might be the output (or an alias of the output) of one of the other `{view}_scatter` ops that you want to remove. Example:
```
  #    b = a[1]
  #    c = b[1]
  #    c_updated = c.add(1)
  #    b_updated = select_scatter(b, c_updated, 0, 1)
  #    a_updated = select_scatter(a, b_updated, 0, 1)
```

I think the high bit here though is that I'm pretty sure it's unnecessary to check for strides or "identical chains of views" here, because it should be enough just to check that the inputs to the scatter ops are in the correct alias sets.

bdhirsh replied:
Nevermind - it looks like like strides / chains of view metadata are still needed.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+
+
+
+def _satisfies_inverse_view(view_inverse_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node]):
```
It's not clear to me what exactly this function does. Top level comment would help.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+
+
+def _satisfies_inverse_view(view_inverse_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node]):
+    for alias in tensor_aliases:
```
Under what circumstances would we return True from this function, AND there are multiple tensor aliases?

bdhirsh replied:
That could technically happen if you have multiple view tensors that are "identical". But I think that if you were to CSE so that we never bother to construct the same view tensor twice, then it would be the case that there's at most 1 tensor in `tensor_aliases` that can satisfy the criteria here.

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+#       e.g. _ = torch.select_scatter(x, mutated_view, args...)
+#       AND the arguments to the view and its scatter operator are the same (args... is equal).
+#   This is effectively undoing the view-inverse logic performed by functionalization.
+def _get_all_later_node_usages(curr_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node], op_index: int, *, only_include_view_inverse_nodes: bool = False) -> Set[torch.fx.node.Node]:
```
So if this function tests if the tensor can be reinplaced, why is the function named get all later node usages

ezyang replied:
It's also not clear what all of the arguments mean

ezyang replied:
This is not just node usages, it is usages of any alias of the node!

bdhirsh replied:
Yes - agreed that was pretty unclear, will beef up the description

bdhirsh replied:
Basically, originally I wanted a function that was just `can_reinplace(...) -> bool`. That function needs to find any candidate `*_scatter*` ops later in the graph that can be removed, in order to figure out if the current node can be replaced. Later on though we need to get any `*_scatter` nodes directly if they exist so that we can remove them from the graph, so I figured I'd have this function directly return those nodes.

So this function really has two modes:
- `only_include_view_inverse_nodes =False`: get all later node usages (and their aliases) of the current self_arg (`curr_node`) that occur later than `op_index` in the graph (where `op_index` is the node we're trying to re-inplace)
- `only_include_view_inverse_nodes` same as the above, but only include the `{view}_scatter` ops

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+    # print(f'node_usages: {", ".join([str(x) for x in n.meta["node_usages"]])}')
+    # if 'view_of' in n.meta:
+    # print(f'view_of: {str(n.meta["view_of"])}')
+    # print()
```
we should setup some permanent logging facilities for this, off by default but easily enable-able

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+
+
+@compatibility(is_backward_compatible=True)
+def reinplace(gm, *sample_args):
```
Is this function valid to run on a non-functional graph?

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+        if node.op == 'call_function':
+            # Step 1: Check to see if this operator has an inplace variant.
+            maybe_inplace_op = _maybe_get_inplace_op(node.target)
+            if maybe_inplace_op is not None:
```
you could avoid indenting here by inverting the condition and continue'ing

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+            if maybe_inplace_op is not None:
+                # This is a proxy check for ensuring that the first argument is "tensor-like"
+                # (This should be the case for all ops with inplace variants in ATen,
+                # although we technically don't have guarantees for custom ops).
```
would be nice to have a test for custom ops. Can probably set it up with python op registration

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+                # If not (same goes for its aliases), then this op is safe to re-in place.
+                # This is a slightly roundabout way to check that there are no later usages of the current self argument.
+                # (later_view_inverse_node_usages corresponds to "view_scatter" nodes that we are allowed to delete)
+                if len(later_node_usages - later_view_inverse_node_usages) == 0:
```
Is it ever the case that `later_view_inverse_node_usages - later_node_usages` is nonzero

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+                        for node_to_update in nodes_to_update:
+                            new_args = []
+                            for arg_idx, a in enumerate(node_to_update.args):
+                                if isinstance(a, torch.fx.node.Node) and a.name == old.name:
```
huh, how come you can't just test for node identity equality

ezyang commented on torch/fx/passes/reinplace.py:
```diff
@@ -0,0 +1,540 @@
+                    all_later_view_inverse_node_usages += later_view_inverse_node_usages
+
+                    # Now that we've replaced b = a.foo() with a.foo_(),
+                    # We need to replace any later usages of "b" with "a"
```
Can you use Node.replace_all_uses_with?

bdhirsh replied:
I ended up fixing a bug, where my `storage_to_nodes: Dict[StorageWeakRef, Set[Node]]` dict wasn't being kept up to date:

When re-inplacing `b = add(a)`, and replacing all usages of `b` with `a`, I actually need to check if there were any view ops later like `x = view(b)` that need to get replaced with `x = view(a)` - and if there are any, update the `storage_to_nodes` mapping afterwards (since now, `x` and `a` are aliased).

I'm not sure if there's a cleaner way to do it :/ But for now, I need to loop over all of the nodes to replace anyway.

ezyang replied:
this kind of finicky metadata handling is one of the reasons why I highly prefer one pass graph transformation algos :)

ezyang reviewed (approved):
very nice work. Let's get this in and take it for a spin on real models

bdhirsh committed:
Update on "add a reinplacing FX pass"

Adds a "reinplacing" FX transform, that goes through an FX graph and tries to convert out-of-place op calls into inplace calls whenever possible.


Followups from this PR include:
- Set up torch bench, and run the whole torchbench suite using AOTAutograd + functionalize + rein placing transforms to surface any issues (this is what I'm currently working on). Right now, I have some basic unit tests just to sanity check that the general logic makes sense.
- Add any missing inplace ops. This is mostly the `*_scatter*` ops, e.g. `diagonal_scatter_`, because these ops will commonly show up an FX graph after running functionalization.

The criteria for when you can swap an op `b = a.add(...)` with `a.add_(...)` is:
(1) An inplace variant of the operator with the same schema needs to exist (`aten.add` -> `aten.add_`)
(2) `a` (**or any of its aliases**) can't be used as an input to any other operators later on in the graph
(3) `a` can't be one of the inputs to the entire graph. It also can't be an **alias** of any of the inputs ***

*** One thing to note: (3) means that we can't technically guarantee that we'll get back **all** memory usage that we lost from functionalization. Functionalization converts input mutations into out-of-place calls, and then adds a `copy_()` to the end of the graph to preserve semantics.

I added logic to handle `copy_()` in this PR because it it's a pretty important optimizations in the context of `functionalization()`: any program that performs input mutations will have a `copy_()` in it after running functionalization.

There are some examples in the test file, but I think staring at an example of where re-inplacing is/isn't allowed to run is helpful:
```
// Before functionalization
def foo(a):
    tmp1 = a.add_(1)
    tmp2 = a.add(2)

// After functionalization
def foo(a)
    tmp1 = a.add(1)
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)

// After re-inplacing
def foo(a)
    // first add() is safe to re-inplace even though a is a program input,
    // because a's data is overwritten later by a copy_()
    tmp1 = a.add_(1)
    // second add() is NOT safe to re-inplace, because:
    // (1) a and tmp1 are aliased. Note that they weren't aliased in the original program,
             but they are now that we've done some re-inplacing.
    // (2) tmp1 is used as an input later in the program
    tmp2 = a.add(2)
    ....
    a.copy_(tmp1)
```





[ghstack-poisoned]

bdhirsh commented:
There are some followups needed, but I'm going to merge this for now and spend some time trying to harden this against torchbench

bdhirsh commented:
@pytorchbot merge

eellison commented:
This is breaking master - see https://github.com/pytorch/pytorch/runs/7548446152?check_suite_focus=true

malfet commented:
@pytorchbot revert -m "broke windows trunk tests, see https://hud.pytorch.org/pytorch/pytorch/commit/3ef7a6921d0c68aba5d1e1e7ebd21b2c89cfc78a" -c nosignal

bdhirsh commented:
Hmm @malfet I ran with `ci/trunk`, and I see the `trunk / win-vs2019-cuda11.6-py3 / test (default, 3, 5, windows.8xlarge.nvidia.gpu)` test now passing. Do you think the failure was flakiness? The error looks pretty obscure:

```
usage_log.txt : The process cannot access the file because it is being used by another process.
[90](https://github.com/pytorch/pytorch/runs/7548446152?check_suite_focus=true#step:11:94)
----------------
[91](https://github.com/pytorch/pytorch/runs/7548446152?check_suite_focus=true#step:11:95)
WARNING: Cannot open 1 file
[92](https://github.com/pytorch/pytorch/runs/7548446152?check_suite_focus=true#step:11:96)
Error: Process completed with exit code 1.
```
